core/src/test/java/com/linecorp/armeria/server/ServerListenerTest.java

```diff
@@ -100,4 +100,146 @@ public void serverStarting(Server server) throws Exception {
                 .hasCauseInstanceOf(IllegalStateException.class)
                 .hasRootCause(ex);
     }
+
+    @Test
+    void testGracefulShutdownPeriodicExecutorWaitIndefinitelyOnServerStopping() throws Exception {
+        final ExecutorService executor = periodicExecutor();
+        final ServerListener sl = ServerListener.builder()
+                                                .stoppingWithExecutor(executor)
+                                                .build();
+
+        final Server server = Server.builder()
+                                    .service("/", (req, ctx) -> HttpResponse.of("Hello!"))
+                                    .serverListener(sl)
+                                    .build();
+        server.start().get();
+        server.stop();
+        Thread.sleep(3000);
+
+        assertThat(executor.isTerminated()).isEqualTo(true);
+    }
+
+    @Test
+    void testGracefulShutdownInterruptibleExecutorWaitIndefinitelyOnServerStopping() throws Exception {
+        final ExecutorService executor = interruptibleExecutor();
+        final ServerListener sl = ServerListener.builder()
+                                                .stoppingWithExecutor(executor)
+                                                .build();
+
+        final Server server = Server.builder()
+                                    .service("/", (req, ctx) -> HttpResponse.of("Hello!"))
+                                    .serverListener(sl)
+                                    .build();
+        server.start().get();
+        server.stop();
+        Thread.sleep(3000);
+
+        assertThat(executor.isTerminated()).isEqualTo(false);
+    }
```

Comment:
For example, this test runs an infinite loop but can be interrupted to stop the `Executor`. It waits infinitely because it does not send an interrupt to the `Executor`, so it is normal behavior that it doesn't terminate no matter how long we wait.

When running this test individually, there is no issue. However, because `Server#stop` waits for `ServerListener` synchronously, the `Server` does not finish, and the test ends with the `Server` still running. As a result, when trying to run the next test, executing `Server#start` will result in the lock not being released, preventing the test from proceeding.

One solution I have considered is modifying the existing `ShutdownSupport#of` method to wait for the termination of the `Executor` asynchronously instead of synchronously. Currently, `ShutdownSupport#of` returns a `CompletableFuture`, but internally, it waits for termination synchronously and returns `UnmodifiableFuture.completedFuture(null)`. By making this code execute asynchronously, it seems that `Server#stop` will finish correctly. (Still, any `Executor` that has not terminated will remain until the all test completes.)

The second approach is to extend the `Executor` used for testing and add an API that forcefully terminates it through means other than interrupts. We can then call this API when the test finishes to properly shut down both the server and the `Executor`.

Reply:
I resolved the issue by making the Executor used in tests AutoCloseable, so that it is closed together when the test ends.